compiler/packages/babel-plugin-react-compiler/src/Optimization/OutlineJsx.ts

```diff
@@ -0,0 +1,405 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import invariant from 'invariant';
+import {Environment} from '../HIR';
+import {
+  BasicBlock,
+  GeneratedSource,
+  HIRFunction,
+  IdentifierId,
+  Instruction,
+  InstructionId,
+  InstructionKind,
+  JsxAttribute,
+  JsxExpression,
+  LoadGlobal,
+  makeBlockId,
+  makeIdentifierName,
+  makeInstructionId,
+  makeType,
+  ObjectProperty,
+  Place,
+  promoteTemporary,
+  promoteTemporaryJsxTag,
+} from '../HIR/HIR';
+import {createTemporaryPlace} from '../HIR/HIRBuilder';
+import {printIdentifier} from '../HIR/PrintHIR';
+import {deadCodeElimination} from './DeadCodeElimination';
+
+export function outlineJSX(fn: HIRFunction): void {
+  const outlinedFns: Array<HIRFunction> = [];
+  outlineJsxImpl(fn, outlinedFns);
+
+  for (const outlinedFn of outlinedFns) {
+    fn.env.outlineFunction(outlinedFn, 'Component');
+  }
+}
+
+type JsxInstruction = Instruction & {value: JsxExpression};
+type LoadGlobalInstruction = Instruction & {value: LoadGlobal};
+type LoadGlobalMap = Map<IdentifierId, LoadGlobalInstruction>;
+
+type State = {
+  jsx: Array<JsxInstruction>;
+  children: Set<IdentifierId>;
+};
+
+function outlineJsxImpl(
+  fn: HIRFunction,
+  outlinedFns: Array<HIRFunction>,
+): void {
+  const globals: LoadGlobalMap = new Map();
+
+  function processAndOutlineJSX(
+    state: State,
+    rewriteInstr: Map<InstructionId, Array<Instruction>>,
+  ): void {
+    if (state.jsx.length <= 1) {
+      return;
+    }
+    const result = process(fn, [...state.jsx].reverse(), globals);
+    if (result) {
+      outlinedFns.push(result.fn);
+      rewriteInstr.set(state.jsx.at(0)!.id, result.instrs);
+    }
+  }
+
+  for (const [, block] of fn.body.blocks) {
+    const rewriteInstr = new Map();
+    let state: State = {
+      jsx: [],
+      children: new Set(),
+    };
+
+    for (let i = block.instructions.length - 1; i >= 0; i--) {
+      const instr = block.instructions[i];
+      const {value, lvalue} = instr;
+      switch (value.kind) {
+        case 'LoadGlobal': {
+          globals.set(lvalue.identifier.id, instr as LoadGlobalInstruction);
+          break;
+        }
+        case 'FunctionExpression': {
+          outlineJsxImpl(value.loweredFunc.func, outlinedFns);
+          break;
+        }
+
+        case 'JsxExpression': {
+          if (!state.children.has(lvalue.identifier.id)) {
+            processAndOutlineJSX(state, rewriteInstr);
+
+            state = {
+              jsx: [],
+              children: new Set(),
+            };
+          }
+          state.jsx.push(instr as JsxInstruction);
+          if (value.children) {
+            for (const child of value.children) {
+              state.children.add(child.identifier.id);
+            }
+          }
+          break;
+        }
```

Comment:
what about other instruction types?

Reply:
this seems to work fine in my testing as we only outline jsx that are immediate children and any mutating instructions are lifted above the jsx so we outline the result correctly, is there an example where this breaks?

Reply:
ah i see what you mean. if a jsx element isn't directly consumed as a child then we reset, and otherwise it's safe to reorder the jsx elements later. cool!